### PR TITLE
egl/android: Defer format detection to gralloc

### PR DESCRIPTION
**Pull request by harishkrupo**
Let gralloc decide the format of the buffer when the
the format is HAL_PIXEL_FORMAT_IMPLEMENTATION_DEFINED

Signed-off-by: Harish Krupo <harish.krupo.kps@intel.com>

### CODE REVIEW

**Comment by tpalli** [comment]
Harish, is there something that requires this? Are we missing some format from the list?

**Comment by harishkrupo** [comment]
Camera Service/Camera Hal requests the creation of an EGLImage with the buffer format HAL_PIXEL_FORMAT_IMPLEMENTATION_DEFINED. This format could be anything based on the camera hardware. Here we use NV12 format. So defer the format identification to gralloc, where it will check the buffer usage bit and return the correct format. In this case HAL_PIXEL_FORMAT_NV12.

**Comment by tpalli** [comment]
This means gralloc needs to know camera hw details (which format to use). Wouldn't it be better to abstract this in Camera HAL and call gralloc with defined format that is suitable for the camera HW used?

**Comment by harishkrupo** [comment]
Not sure about the camera HAL but people at chromium are also trying to do something like this to handle the IMPLEMENTATION_DEFINED pixel format passed to both gralloc and mesa. 
[chromium WIP patch](https://chromium-review.googlesource.com/c/487946/2/src/egl/drivers/dri2/platform_android.c)

**Comment by kalyankondapally** [comment]
@harishkrupo Can we get this rebased.

**Comment by harishkrupo** [comment]
@kalyankondapally Rebased.

File: src/egl/drivers/dri2/platform_android.c [hunk 1]
```diff
@@ -222,6 +224,28 @@ droid_window_dequeue_buffer(struct dri2_egl_surface *dri2_surf)
...
+   int format;
+   int err;
+
+   if (buf->format != HAL_PIXEL_FORMAT_IMPLEMENTATION_DEFINED)
+      return buf->format;
+
+   if (!dri2_dpy->gralloc->perform)
+      return -1;
+
+   err = dri2_dpy->gralloc->perform(dri2_dpy->gralloc,
```
Comment (kalyankondapally) [inline src/egl/drivers/dri2/platform_android.c:240]: wondering what we need to do for Gralloc 1.0 ?

Comment (harishkrupo) [reply src/egl/drivers/dri2/platform_android.c:240]: Looks like Gralloc 1.0's perform call has support for the GRALLOC_DRM_GET_FORMAT operation. So I don't think any change is required, still, I will test it and report back.

Comment (kalyankondapally) [reply src/egl/drivers/dri2/platform_android.c:240]: There is no perform call for Gralloc 1.0

Comment (harishkrupo) [reply src/egl/drivers/dri2/platform_android.c:240]: gralloc1.0-rebase is not the correct branch?

Comment (kalyankondapally) [reply src/egl/drivers/dri2/platform_android.c:240]: It is you can find the API here: https://github.com/01org/minigbm/blob/gralloc1.0-rebase/cros_gralloc/gralloc1/cros_gralloc1_module.cc

Comment (harishkrupo) [reply src/egl/drivers/dri2/platform_android.c:240]: oops, my bad checked gralloc0.

**Comment by tpalli** [comment]
In the chromium WIP patch above it is clearly marked that this is a temporary hack ... maybe should discuss with Tomasz to see if there's better solution coming up.

**Comment by harishkrupo** [comment]
Anyways, this patch will not be suitable once we switch to Gralloc 1.0 as it depends on the perform method in gralloc, which is not available in 1.0

**Comment by tpalli** [comment]
with gralloc1 there is GRALLOC1_PFN_GET_FORMAT that could be used in similar way, I'll ping Tomasz if there's something better coming

**Comment by kalyankondapally** [comment]
I think we could use that

**Comment by kalyankondapally** [comment]
@tpalli ANy objections to merge this and add gralloc1.0 support as part of that PR which is going through reviews

**Comment by tpalli** [comment]
No objections, please do. There is discussion going on about implementing a small library interface that would hide gralloc0/gralloc1 but this will take some time to materialize so let's use now what we have!

**Comment by kalyankondapally** [comment]
merged

**Pull request closed without merging**